File: killrvideo-service-videocatalog/src/main/java/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcMapper.java
package com.killrvideo.service.video.grpc;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import com.google.common.collect.Sets;
import com.google.protobuf.Timestamp;
import com.killrvideo.dse.dto.CustomPagingState;
import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.service.video.dto.LatestVideo;
import com.killrvideo.service.video.dto.LatestVideosPage;
import com.killrvideo.service.video.dto.UserVideo;

import com.killrvideo.dse.dto.Video;
import com.killrvideo.service.video.request.GetLatestVideoPreviewsRequestData;
import com.killrvideo.service.video.request.GetUserVideoPreviewsRequestData;
import com.killrvideo.utils.GrpcMappingUtils;
import killrvideo.common.CommonTypes;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.GetLatestVideoPreviewsResponse;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.GetVideoResponse;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.SubmitYouTubeVideoRequest;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.VideoLocationType;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.VideoPreview;
import killrvideo.video_catalog.events.VideoCatalogEvents;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Component;

import static com.killrvideo.utils.GrpcMappingUtils.*;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

/**
 * Utility mapping GRPC.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
public class VideoCatalogServiceGrpcMapper {

    public Video mapSubmitYouTubeVideoRequestAsVideo(SubmitYouTubeVideoRequest request) {
        Video targetVideo = new Video();
        targetVideo.setVideoid(fromUuid(request.getVideoId()));
        targetVideo.setUserid(fromUuid(request.getUserId()));
        targetVideo.setName(request.getName());
        targetVideo.setLocation(request.getYouTubeVideoId());
        targetVideo.setDescription(request.getDescription());
        targetVideo.setPreviewImageLocation("//img.youtube.com/vi/" + targetVideo.getLocation() + "/hqdefault.jpg");
        targetVideo.setTags(Sets.newHashSet(request.getTagsList().iterator()));
        targetVideo.setLocationType(VideoLocationType.YOUTUBE.ordinal());
        return targetVideo;
    }

    public GetLatestVideoPreviewsResponse mapLatestVideoToGrpcResponse(LatestVideosPage returnedPage) {
        return GetLatestVideoPreviewsResponse.newBuilder()
                .addAllVideoPreviews(
                        returnedPage.getListOfPreview().stream()
                                .map(this::mapLatestVideotoVideoPreview)
                                .collect(Collectors.toList()))
                .setPagingState(returnedPage.getNextPageState())
                .build();
    }

    /**
     * Mapping to GRPC generated classes.
     */
    private VideoPreview mapLatestVideotoVideoPreview(LatestVideo lv) {
        return VideoPreview.newBuilder()
                .setAddedDate(instantToTimeStamp(lv.getAddedDate()))
                .setName(lv.getName())
                .setPreviewImageLocation(Optional.ofNullable(lv.getPreviewImageLocation()).orElse("N/A"))
                .setUserId(uuidToUuid(lv.getUserid()))
                .setVideoId(uuidToUuid(lv.getVideoid()))
                .build();
    }

    /**
     * Mapping to generated GPRC beans.
     */
    private VideoPreview mapFromUserVideotoVideoPreview(UserVideo v) {
        return VideoPreview.newBuilder()
                .setAddedDate(instantToTimeStamp(v.getAddedDate()))
                .setName(v.getName())
                .setPreviewImageLocation(Optional.ofNullable(v.getPreviewImageLocation()).orElse("N/A"))
                .setUserId(uuidToUuid(v.getUserid()))
                .setVideoId(uuidToUuid(v.getVideoid()))
                .build();
    }

    /**
     * Mapping to generated GPRC beans (Full detailed)
     */
    public GetVideoResponse mapFromVideotoVideoResponse(Video v) {
        // Check to see if any tags exist, if not, ensure to send an empty set instead of null
        if (CollectionUtils.isEmpty(v.getTags())) {
            v.setTags(Collections.emptySet());
        }
        return GetVideoResponse.newBuilder()
                .setAddedDate(instantToTimeStamp(v.getAddedDate()))
                .setDescription(v.getDescription())
                .setLocation(v.getLocation())
                .setLocationType(VideoLocationType.forNumber(v.getLocationType()))
                .setName(v.getName())
                .setUserId(uuidToUuid(v.getUserid()))
                .setVideoId(uuidToUuid(v.getVideoid()))
                .addAllTags(v.getTags())
                .build();
    }

    public VideoCatalogEvents.YouTubeVideoAdded createYouTubeVideoAddedEvent(Video video) {
        return VideoCatalogEvents.YouTubeVideoAdded.newBuilder()
                .setAddedDate(Timestamp.newBuilder().build())
                .setDescription(video.getDescription())
                .setLocation(video.getLocation())
                .setName(video.getName())
                .setPreviewImageLocation(video.getPreviewImageLocation())
                .setUserId(uuidToUuid(video.getUserid()))
                .setVideoId(uuidToUuid(video.getVideoid()))
                .build();
    }

    public GetLatestVideoPreviewsRequestData parseGetLatestVideoPreviewsRequest(
            GetLatestVideoPreviewsRequest grpcReq,
            Supplier<CustomPagingState> firstCustomPagingStateFactory) {
        CustomPagingState pageState =
                CustomPagingState.deserialize(grpcReq.getPagingState())
                        .orElse(firstCustomPagingStateFactory.get());
        int pageSize = grpcReq.getPageSize();
        final Optional<Instant> startDate = Optional.of(grpcReq.getStartingAddedDate())
                .filter(x -> isNotBlank(x.toString()))
                .map(GrpcMappingUtils::timestampToInstant);
        final Optional<UUID> startVideoId = Optional.of(grpcReq.getStartingVideoId())
                .filter(x -> isNotBlank(x.toString()))
                .map(GrpcMappingUtils::fromUuid);

        return new GetLatestVideoPreviewsRequestData(pageState, pageSize, startDate, startVideoId);
    }

    public GetVideoPreviewsResponse mapToGetVideoPreviewsResponse(List<Video> videos) {
        final GetVideoPreviewsResponse.Builder builder = GetVideoPreviewsResponse.newBuilder();
        videos.stream()
                .map(this::mapFromVideotoVideoPreview)
                .forEach(builder::addVideoPreviews);
        return builder.build();
    }

    /**
     * Mapping to generated GPRC beans.
     */
    private VideoPreview mapFromVideotoVideoPreview(Video v) {
        return VideoPreview.newBuilder()
                .setAddedDate(instantToTimeStamp(v.getAddedDate()))
                .setName(v.getName())
                .setPreviewImageLocation(Optional.ofNullable(v.getPreviewImageLocation()).orElse("N/A"))
                .setUserId(uuidToUuid(v.getUserid()))
                .setVideoId(uuidToUuid(v.getVideoid()))
                .build();
    }

    public GetUserVideoPreviewsRequestData parseGetUserVideoPreviewsRequest(GetUserVideoPreviewsRequest grpcReq) {
        final UUID userId = fromUuid(grpcReq.getUserId());
        final Optional<UUID> startingVideoId = Optional
                .of(grpcReq.getStartingVideoId())
                .filter(uuid -> isNotBlank(uuid.getValue()))
                .map(GrpcMappingUtils::fromUuid);
        final Optional<Instant> startingAddedDate = Optional
                .of(grpcReq.getStartingAddedDate())
                .map(GrpcMappingUtils::timestampToInstant);
        final Optional<Integer> pagingSize =
                Optional.of(grpcReq.getPageSize());
        final Optional<String> pagingState =
                Optional.of(grpcReq.getPagingState())
                        .filter(StringUtils::isNotBlank);

        return new GetUserVideoPreviewsRequestData(
                userId, startingVideoId, startingAddedDate, pagingSize, pagingState
        );
    }

    public GetUserVideoPreviewsResponse mapToGetUserVideoPreviewsResponse(ResultListPage<UserVideo> resultPage, UUID userId) {
        CommonTypes.Uuid userGrpcUUID = GrpcMappingUtils.uuidToUuid(userId);
        final GetUserVideoPreviewsResponse.Builder builder = GetUserVideoPreviewsResponse.newBuilder().setUserId(userGrpcUUID);
        resultPage.getResults().stream()
                .map(this::mapFromUserVideotoVideoPreview)
                .forEach(builder::addVideoPreviews);
        resultPage.getPagingState().ifPresent(builder::setPagingState);
        return builder.build();
    }
}

File: killrvideo-service-videocatalog/src/main/kotlin/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcMapper.kt
package com.killrvideo.service.video.grpc

import com.google.common.collect.Sets
import com.google.protobuf.Timestamp
import com.killrvideo.dse.dto.CustomPagingState
import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.service.video.dto.LatestVideo
import com.killrvideo.service.video.dto.LatestVideosPage
import com.killrvideo.service.video.dto.UserVideo
import com.killrvideo.service.video.dto.Video
import com.killrvideo.service.video.request.GetLatestVideoPreviewsRequestData
import com.killrvideo.service.video.request.GetUserVideoPreviewsRequestData
import com.killrvideo.utils.GrpcMappingUtils.*
import killrvideo.video_catalog.*
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*
import killrvideo.video_catalog.events.VideoCatalogEvents.YouTubeVideoAdded
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.springframework.stereotype.Component
import java.util.*
import java.util.function.Supplier
import java.util.stream.Collectors

/**
 * Utility mapping GRPC.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
class VideoCatalogServiceGrpcMapper {
    object SubmitYouTubeVideoRequestExtensions {
        fun SubmitYouTubeVideoRequest.parse(): Video =
            Video(
                videoid = fromUuid(this.videoId),
                userid = fromUuid(this.userId),
                name = this.name,
                description = this.description,
                location = this.youTubeVideoId,
                locationType = VideoLocationType.YOUTUBE.ordinal,
                previewImageLocation = "//img.youtube.com/vi/${this.youTubeVideoId}/hqdefault.jpg",
                tags = Sets.newHashSet(this.tagsList.iterator()),
                addedDate = null
            )
    }

    object GetUserVideoPreviewsRequestExtensions {
        fun GetUserVideoPreviewsRequest.parse() =
            GetUserVideoPreviewsRequestData(
                userId = fromUuid(this.userId),
                startingVideoId = if (this.hasStartingVideoId() && isNotBlank(this.startingVideoId.value)) {
                    fromUuid(this.startingVideoId)
                } else null,
                startingAddedDate = if (this.hasStartingAddedDate())
                    timestampToInstant(this.startingAddedDate)
                else null,
                pagingSize = if (this.pageSize > 0) this.pageSize else null,
                pagingState = if (isNotBlank(this.pagingState)) this.pagingState else null
            )
    }

    object GetLatestVideoPreviewsRequestExtensions {
        fun GetLatestVideoPreviewsRequest.parse(firstCustomPagingStateFactory: Supplier<CustomPagingState>):
                GetLatestVideoPreviewsRequestData =
            GetLatestVideoPreviewsRequestData(
                pageState = if (isNotBlank(this.pagingState))
                    CustomPagingState.deserialize(this.pagingState).orElse(firstCustomPagingStateFactory.get())
                else
                    firstCustomPagingStateFactory.get(),
                pageSize = this.pageSize,
                startDate = if (this.hasStartingAddedDate()) timestampToInstant(this.startingAddedDate) else null,
                startVideoId = if (this.hasStartingVideoId() && isNotBlank(this.startingVideoId.value))
                    fromUuid(this.startingVideoId) else null
            )
    }

    fun mapLatestVideoToGrpcResponse(returnedPage: LatestVideosPage): GetLatestVideoPreviewsResponse =
        getLatestVideoPreviewsResponse {
            videoPreviews.addAll(
                returnedPage.listOfPreview.stream()
                    .map { lv: LatestVideo -> mapLatestVideotoVideoPreview(lv) }
                    .collect(Collectors.toList())
            )
            returnedPage.nextPageState?.let { pagingState = it }
        }

    /**
     * Mapping to GRPC generated classes.
     */
    private fun mapLatestVideotoVideoPreview(v: LatestVideo): VideoPreview =
        videoPreview {
            v.addedDate?.let { addedDate = instantToTimeStamp(it) }
            v.name?.let { name = it }
            previewImageLocation = v.previewImageLocation ?: "N/A"
            v.userid?.let { userId = uuidToUuid(it) }
            v.videoid?.let { videoId = uuidToUuid(it) }
        }

    /**
     * Mapping to generated GPRC beans.
     */
    private fun mapFromUserVideotoVideoPreview(v: UserVideo): VideoPreview =
        videoPreview {
            v.addedDate?.let { addedDate = instantToTimeStamp(it) }
            v.name?.let { name = it }
            previewImageLocation = v.previewImageLocation ?: "N/A"
            v.userid?.let { userId = uuidToUuid(it) }
            v.videoid?.let { videoId = uuidToUuid(it) }
        }

    /**
     * Mapping to generated GPRC beans (Full detailed)
     */
    fun mapFromVideotoVideoResponse(v: Video): GetVideoResponse =
        // Check to see if any tags exist, if not, ensure to send an empty set instead of null
        getVideoResponse {
            v.addedDate?.let { addedDate = instantToTimeStamp(it) }
            v.description?.let { description = it }
            v.location?.let { location = it }
            v.locationType?.let { locationType = VideoLocationType.forNumber(it) }
            v.name?.let { name = it }
            v.userid?.let { userId = uuidToUuid(it) }
            v.videoid?.let { videoId = uuidToUuid(it) }
            v.tags?.let { tags.addAll(it) }
        }

    fun createYouTubeVideoAddedEvent(video: Video): YouTubeVideoAdded {
        val builder = YouTubeVideoAdded.newBuilder()
            .setAddedDate(Timestamp.newBuilder().build())

        video.description?.let { builder.setDescription(it) }
        video.location?.let { builder.setLocation(it) }
        video.name?.let { builder.setName(it) }
        video.previewImageLocation?.let { builder.setPreviewImageLocation(it) }
        video.userid?.let { builder.setUserId(uuidToUuid(it)) }
        video.videoid?.let { builder.setVideoId(uuidToUuid(it)) }

        return builder.build()
    }

    fun mapToGetVideoPreviewsResponse(videos: List<Video>): GetVideoPreviewsResponse =
        getVideoPreviewsResponse {
            videoPreviews.addAll(
                videos.stream().map { mapFromVideotoVideoPreview(it) }.collect(Collectors.toList())
            )
        }

    /**
     * Mapping to generated GPRC beans.
     */
    private fun mapFromVideotoVideoPreview(v: Video): VideoPreview =
        videoPreview {
            v.addedDate?.let { addedDate = instantToTimeStamp(it) }
            v.name?.let { name = it }
            previewImageLocation = v.previewImageLocation ?: "N/A"
            v.userid?.let { userId = uuidToUuid(it) }
            v.videoid?.let { videoId = uuidToUuid(it) }
        }

    fun mapToGetUserVideoPreviewsResponse(
        resultPage: ResultListPage<UserVideo?>,
        userid: UUID
    ): GetUserVideoPreviewsResponse =
        getUserVideoPreviewsResponse {
            userId = uuidToUuid(userid)
            resultPage.pagingState.ifPresent { pagingState = it }
            resultPage.results.stream()
                .filter { x -> x != null }
                .map { mapFromUserVideotoVideoPreview(it!!) }
                .forEach { videoPreviews.add(it) }
        }
}
